python-package/lightgbm/sklearn.py

```diff
@@ -1067,6 +1137,21 @@ def n_features_in_(self) -> int:
             raise LGBMNotFittedError("No n_features_in found. Need to call fit beforehand.")
         return self._n_features_in
 
+    @n_features_in_.setter
```

Comment:
If you pass `reset=True` to `sklearn.utils.validation.validate_data()`, it will try to:

* set `estimator.n_features_in_` ([code link](https://github.com/scikit-learn/scikit-learn/blob/545d99e0fd1de69b317496c77bd5c92a46cd1a9e/sklearn/utils/validation.py#L2757-L2759))
* delete `estimator.feature_names_in_` ([code link](https://github.com/scikit-learn/scikit-learn/blob/545d99e0fd1de69b317496c77bd5c92a46cd1a9e/sklearn/utils/validation.py#L2653-L2660))

We want the "set `estimator.n_features_in_`" behavior, because without it we have to manually set `estimator.n_features_in_` in `fit()`.

Doing that requires determining the number of features in `X`, which requires either re-implementing something like `sklearn.utils.validation._num_features()` (as I originally tried to do) or just calling that function directly. But that function can't safely be called directly before calling `check_array()`, because it raises a `TypeError` on 1-D inputs, which violates the `check_fit1d` estimator check ([code link](https://github.com/scikit-learn/scikit-learn/blob/545d99e0fd1de69b317496c77bd5c92a46cd1a9e/sklearn/utils/validation.py#L1074-L1091)).

So here, I'm proposing that we do the following:

* add a setter for `n_features_in_` and a deleter for `feature_names_in_`
* pass `reset=True` at `fit()` time to `validate_data()`
* modify the pre-1.6 implementation of `validate_data()` in `compat.py` to match